File: solutions/elixir/all-your-base/3/lib/all_your_base.ex
defmodule AllYourBase do
  @doc """
  Given a number in input base, represented as a sequence of digits, converts it to output base,
  or returns an error tuple if either of the bases are less than 2
  """

  @spec convert(list, integer, integer) :: {:ok, list} | {:error, String.t()}
  def convert(_, _, output_base) when output_base < 2, do: {:error, "output base must be >= 2"}
  def convert(_, input_base, _) when input_base < 2, do: {:error, "input base must be >= 2"}

  def convert(digits, input_base, output_base) do
    if !Enum.all?(digits, &(0 <= &1 and &1 < input_base)) do
      {:error, "all digits must be >= 0 and < input base"}
    else
      {:ok,
       digits
       |> to_int(input_base, 0)
       |> to_base(output_base, [])}
    end
  end

  defp to_int([], _, acc), do: acc
  defp to_int([d | t], b, acc), do: to_int(t, b, acc * b + d)

  defp to_base(n, b, acc) when n < b, do: [n | acc]
  defp to_base(n, b, acc), do: to_base(div(n, b), b, [rem(n, b) | acc])
end

File: solutions/cpp/all-your-base/3/all_your_base.cpp
#include "all_your_base.h"

#include <algorithm>
#include <numeric>
#include <cmath>

namespace all_your_base {
    using namespace std;

    unsigned int to_int(const vector<unsigned int> &digits, int base) {
        unsigned int e = 0;
        return accumulate(digits.rbegin(), digits.rend(), 0, [&e, base](unsigned int acc, unsigned int digit) {
            return acc + (pow(base, e++) * digit);
        });
    }

    auto to_base(unsigned int i, int base) {
        vector<unsigned int> c;
        for (; i > 0; i /= base) {
            c.push_back(i % base);
        }
        ranges::reverse(c);
        return c;
    }

    vector<unsigned int> convert(int input_base, const vector<unsigned int> &digits, int output_base) {
        if (input_base < 2) throw invalid_argument("input_base < 2");
        if (output_base < 2) throw invalid_argument("output_base < 2");

        const auto predicate = [input_base](int x) { return x < 0 || x >= input_base; };
        if (any_of(digits.begin(), digits.end(), predicate)) throw invalid_argument("invalid digit");

        return to_base(
                to_int(digits, input_base),
                output_base);
    }
}  // namespace all_your_base
